pkg/operator/encryption/state.go

```diff
@@ -43,175 +45,241 @@ func getCurrentEncryptionConfig(secrets corev1client.SecretInterface, revision s
 				break outer
 			}
 		}
 	}
-	// if our read secrets aren't all the same, the first thing to do is to force all the read secrets and wait for stability
-	// TODO if the operand namespace gets deleted, this code causes us to go through a cycle of identity as the write key
 	if !allReadSecretsAsExpected {
 		klog.V(4).Infof("not all read secrets in sync")
-		for gr := range encryptionState {
-			grState := encryptionState[gr]
-			grState.readSecrets = encryptionSecrets
-			encryptionState[gr] = grState
+		for gr := range desiredEncryptionState {
+			grState := desiredEncryptionState[gr]
+			// TODO: here we might lose read key if the "wrong" secrets are deleted
+			grState.readSecrets = expectedReadSecrets
+			desiredEncryptionState[gr] = grState
 		}
-		return encryptionState
+		return desiredEncryptionState
 	}
 
-	// we do not have any keys yet, so wait until we do
-	// the code after this point assumes at least one secret
-	if len(encryptionSecrets) == 0 {
-		klog.V(4).Infof("no encryption secrets found")
-		return encryptionState
-	}
-
-	// the first secret holds the write key
+	//
+	// STEP 3: with consistent read-keys, verify first read-key is write-key. If not, set write-key and wait for stability.
+	//
 	writeSecret := encryptionSecrets[0]
-
-	// we have consistent and completely current read secrets, the next step is determining a consistent write key.
-	// To do this, we will choose the most current write key and use that to write the data
 	allWriteSecretsAsExpected := true
-	for gr, grState := range encryptionState {
+	for gr, grState := range desiredEncryptionState {
 		if grState.writeSecret == nil || grState.writeSecret.Name != writeSecret.Name {
 			allWriteSecretsAsExpected = false
 			klog.V(4).Infof("%s does not have write secret %s", gr, writeSecret.Name)
 			break
 		}
 	}
-	// if our write secrets aren't all the same, update all the write secrets and wait for stability.
-	// We can move write keys before all the data has been migrated
 	if !allWriteSecretsAsExpected {
 		klog.V(4).Infof("not all write secrets in sync")
-		for gr := range encryptionState {
-			grState := encryptionState[gr]
+		for gr := range desiredEncryptionState {
+			grState := desiredEncryptionState[gr]
 			grState.writeSecret = writeSecret
-			encryptionState[gr] = grState
+			desiredEncryptionState[gr] = grState
 		}
-		return encryptionState
+		return desiredEncryptionState
 	}
 
-	// at this point all of our read and write secrets are the same.  We need to inspect the write secret to
-	// ensure that it has all the expect resources listed as having been migrated.  If this is true, then we
-	// can prune the read secrets down to a single entry.
+	//
+	// STEP 4: with consistent read-keys and write-keys, remove every read-key other than the write-key.
+	//
+	// Note: because read-keys are consistent, currentlyEncryptedGRs equals toBeEncryptedGRs
+	allMigrated, _, reason := migratedFor(currentlyEncryptedGRs, writeSecret)
```

Comment:
is this needed? I think that `expectedReadSecrets` will hold only `writeSecret` when read-keys are consistent.

Reply:
but the write key could be set without having finished the migration.